ibis/sql/postgres/udf/api.py

```diff
@@ -0,0 +1,218 @@
+    """
+    definition = next(_udf_name_cache[name])
+    external_name = '{}_{:d}'.format(name, definition)
+    return type(external_name, (PostgresUDFNode,), fields)
+
+
+def existing_udf(name,
+                 input_types,
+                 output_type,
+                 schema=None,
+                 parameter_names=None):
+    """Create a ibis function that refers to an existing Postgres UDF already
+    defined in database
+
+    Parameters
+    ----------
+    name: str
+    input_types : List[DataType]
+    output_type : DataType
+    schema: str - optionally specify the schema that the UDF is defined in
+    parameter_names: str - give names to the arguments of the UDF
+
+    Returns
+    -------
+    wrapper : Callable
+        The wrapped function
+    """
+    if parameter_names is None:
+        parameter_names = ['v{}'.format(i) for i in range(len(input_types))]
+    else:
+        assert len(input_types) == len(parameter_names)
+
+    udf_node_fields = collections.OrderedDict([
+        (name, Arg(rlz.value(type_)))
+        for name, type_ in zip(parameter_names, input_types)
+    ] + [
+        (
+            'output_type',
+            lambda self, output_type=output_type: rlz.shape_like(
+                self.args, dtype=output_type
+            )
+        )
+    ])
+    udf_node_fields['resolve_name'] = lambda self: name
+
+    udf_node = create_udf_node(name, udf_node_fields)
+
+    def _translate_udf(t, expr):
+        func_obj = sqlalchemy.func
+        if schema is not None:
+            func_obj = getattr(func_obj, schema)
+        func_obj = getattr(func_obj, name)
+
+        sa_args = [t.translate(arg) for arg in expr.op().args]
+
+        return func_obj(*sa_args)
+
+    add_operation(udf_node, _translate_udf)
+
+    def wrapped(*args, **kwargs):
+        node = udf_node(*args, **kwargs)
+        return node.to_expr()
+
+    return wrapped
+
+
+def func_to_udf(conn,
+                python_func,
+                in_types=None,
+                out_type=None,
+                schema=None,
+                name=None,
+                overwrite=False):
+    """Defines a UDF in the database
+
+    Parameters
+    ----------
+    conn: sqlalchemy engine
+    python_func: python function
+    in_types: List[DataType]; if left None, will try to infer datatypes from
+    function signature
+    out_type : DataType
+    schema: str - optionally specify the schema in which to define the UDF
+    name: str - name for the UDF to be defined in database
+    overwrite: bool - replace UDF in database if already exists
+
+    Returns
+    -------
+    wrapper : Callable
+        The ibis UDF object as a wrapped function
+    """
+    if name is None:
+        internal_name = python_func.__name__
+    else:
+        internal_name = name
+    signature = inspect.signature(python_func)
+    parameter_names = signature.parameters.keys()
+    if in_types is None:
+        raise NotImplementedError('inferring in_types not implemented')
```

Comment:
Curious what it means to infer the input types here?

Reply:
Now that Python 3 supports type annotations in function signatures, my plan is to add logic that infer the ibis types from the Python types in the annotations. See example below for getting the Python types from a signature. 

```
import inspect

def total(price: float, quantity: int) -> float:
    return price * quantity

sig = inspect.signature(total)

price_pytype = sig.parameters['price'].annotation
return_pytype = sig.return_annotation
```

If any ibis types are explicitly passed in, those would take precedence. But if none are passed in yet the function has type annotations, it could map those python types to ibis types and use the inferred ones. 

If they were annotated with ibis types already we could skip inference. What I mean by inference is that there's already a [mapping between Python and Postgres types](https://www.postgresql.org/docs/10/plpython-data.html) defined in the PL/Python extension itself, so we can simply use that specification to do the inference.

Reply:
Perhaps I should just remove the placeholders for inferring types since it's not yet implemented anyway. That could be a subsequent issue and pull request, and it would keep the code looking cleaner in the meantime.

Reply:
I just removed the placeholders.